src/lib.rs

```diff
@@ -794,13 +832,47 @@ impl IntoIter {
         FilterEntry { it: self, predicate: predicate }
     }
 
+    #[cfg(unix)]
+    fn is_same_file_system(&self, dent: &DirEntry) -> Result<bool> {
+        use std::os::unix::fs::MetadataExt;
+        let metadata = dent.metadata()?;
+        // We cannot take the error from self since its borrowed, so we make a new error.
+        // If the root_device is unknown, every entry will fail.
+        match self.opts.root_device {
+            Some(Ok(t)) => Ok(metadata.dev() == t),
+            _ => Err(Error::from_entry(dent,std::io::Error::new(
+                std::io::ErrorKind::Other,
+                "Root device id was never identified"))),
+        }
+    }
+
+    #[cfg(windows)]
+    fn is_same_file_system(&self, dent: &DirEntry) -> Result<bool> {
+        let by_handle_file_info = windows::windows_file_handle_info(&dent.path)
+            .map_err(|err| { Error::from_entry(dent, err) })?;
+        // We cannot take the error from self since it's borrowed, so we make a new error.
+        // If the root_device is unknown, every entry will fail.
```

Comment:
This is really unfortunate. We really should surface the original error. Can you please brainstorm ways to make that happen?